pkg/redirectpolicy/redirectpolicy.go

```diff
@@ -0,0 +1,173 @@
+//  Copyright 2020 Authors of Cilium
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package redirectpolicy
+
+import (
+	"fmt"
+	"github.com/cilium/cilium/pkg/k8s"
+	v2 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2"
+	k8sUtils "github.com/cilium/cilium/pkg/k8s/utils"
+	"github.com/cilium/cilium/pkg/loadbalancer"
+	"github.com/cilium/cilium/pkg/policy/api"
+	"k8s.io/apimachinery/pkg/types"
+	"net"
+	"strconv"
+)
+
+type frontend = loadbalancer.L3n4Addr
+
+type backend = loadbalancer.L3n4Addr
+
+// Policy ID that includes its name and namespace
+type policyID = string
+
+func (config LRPConfig) getPolicyID() policyID {
+	return config.name + config.namespace
+}
+
+func Parse(clrp *v2.CiliumLocalRedirectPolicy, sanitize bool) (LRPConfig, error) {
+	name := clrp.ObjectMeta.Name
+	if name == "" {
+		return LRPConfig{},
+			fmt.Errorf("CiliumLocalRedirectPolicy must have a name")
+	}
+
+	namespace := k8sUtils.ExtractNamespace(&clrp.ObjectMeta)
+	// if namespace == "" {
+	// TODO CCLRP
+	// 	return cclrp.Parse()
+	// }
+
+	if sanitize {
+		return getSanitizedLRPConfig(name, namespace, clrp.UID, clrp.Spec)
+	} else {
+		return LRPConfig{
+			name:      name,
+			namespace: namespace,
+		}, nil
+	}
+}
+
+func getSanitizedLRPConfig(name, namespace string, uid types.UID,
+	spec v2.CiliumLocalRedirectPolicySpec) (LRPConfig, error) {
+
+	var (
+		addrMatcher  = spec.RedirectFrom.AddressMatcher
+		svcMatcher   = spec.RedirectFrom.ServiceMatcher
+		redirectTo   = spec.RedirectTo
+		k8sSvc       *k8s.ServiceID
+		svcPortProto api.PortProtocol
+		bePortProto  api.PortProtocol
+		fe           *loadbalancer.L3n4Addr
+	)
+
+	if addrMatcher == nil && svcMatcher == nil {
+		return LRPConfig{}, fmt.Errorf("both address and service" +
+			" matchers can not be empty")
+	}
+
+	if addrMatcher != nil && svcMatcher != nil {
+		return LRPConfig{}, fmt.Errorf("both address and service" +
+			" matchers can not be specified")
+	}
+
+	if addrMatcher != nil {
+		// LRP specifies IP and L4 port tuple config for traffic that needs to be redirected.
+		ip := net.ParseIP(addrMatcher.IP)
+		if ip == nil {
+			return LRPConfig{}, fmt.Errorf("invalid address matcher IP %s",
```

Comment:
Nit: errors should be formatted as `%v`

Reply:
+v is used for printing structs [1]. `%s` seems to be just fine for printing strings, checked other files as well.

[1] https://golang.org/pkg/fmt/

Reply:
For new code, we prefer moving towards either `%v` or `%w` because it allows us to preserve the error when it's passed up the stack. When it is formatted with `%s`, it loses its "context" because it just becomes a "new" error with a longer string. Rather, we can preserve the error and that opens up the possibility to leverage `Unwrap` and friends.

Just for a bit of background on existing code using `%s`, yes we do have code still using `%s`, and that should be looked into eventually. Changing them now _may_ break the callers who depend on the errors in this format. I'm not sure if this is actually concrete problem, but theoretically it could exist.

Reply:
Thanks for providing the background. I'll update the couple of places where I'm using `%s`.